devito/passes/clusters/factorization.py

```diff
@@ -72,6 +75,17 @@ def collect_special(expr):
         else:
             terms.append(i)
 
+    # Any factorization possible?
+    if not any(len(i) > 1 for i in [w_funcs, w_pows, w_coeffs]):
+        # `evaluate=True` below guarantees de-nesting of Adds
+        # For example:
+        # `args[0] = ((0.1*(a + b)) + (0.2*(c + d)))`
+        # `args[1] = ((0.1*(e + f)) + (0.2*(g + h)))`
+        # ->
+        # `expr = (0.1*(a + b) + 0.2*(c + d) + 0.1*(e + f) + 0.2*(g + h))`
+        # Which is then further factorizable by `collect_const`
+        return reuse_if_untouched(expr, args, evaluate=True)
```

Comment:
it's only triggered if any of the args has changed, otherwise `expr` is returned. Yes, this is necessary or we'll miss factorization opportunities down the line